app/src/studio/gatt_rpc_transport.c

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (c) 2024 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/init.h>
+#include <sys/types.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/kernel.h>
+#include <zephyr/bluetooth/gatt.h>
+#include <zephyr/sys/ring_buffer.h>
+
+#include <zmk/ble.h>
+#include <zmk/event_manager.h>
+#include <zmk/events/ble_active_profile_changed.h>
+#include <zmk/studio/rpc.h>
+
+#include "uuid.h"
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_DECLARE(zmk_studio, CONFIG_ZMK_STUDIO_LOG_LEVEL);
+
+static bool handling_rx = false;
+
+static atomic_t notify_size;
+
+static void rpc_ccc_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value) {
+    ARG_UNUSED(attr);
+
+    bool notif_enabled = (value == BT_GATT_CCC_INDICATE);
+
+    LOG_INF("RPC Notifications %s", notif_enabled ? "enabled" : "disabled");
+
+#if CONFIG_ZMK_STUDIO_TRANSPORT_BLE_PREF_LATENCY < CONFIG_BT_PERIPHERAL_PREF_LATENCY
+    struct bt_conn *conn = zmk_ble_active_profile_conn();
+    if (conn) {
+        uint8_t latency = notif_enabled ? CONFIG_ZMK_STUDIO_TRANSPORT_BLE_PREF_LATENCY
+                                        : CONFIG_BT_PERIPHERAL_PREF_LATENCY;
+
+        int ret = bt_conn_le_param_update(
+            conn,
+            BT_LE_CONN_PARAM(CONFIG_BT_PERIPHERAL_PREF_MIN_INT, CONFIG_BT_PERIPHERAL_PREF_MAX_INT,
+                             latency, CONFIG_BT_PERIPHERAL_PREF_TIMEOUT));
+        if (ret < 0) {
+            LOG_WRN("Failed to request lower latency while studio is active (%d)", ret);
+        }
+
+        bt_conn_unref(conn);
+    }
+#endif
+}
+
+static ssize_t read_rpc_resp(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+                             uint16_t len, uint16_t offset) {
+
+    LOG_DBG("Read response for length %d at offset %d", len, offset);
+    return 0;
+}
+
+static ssize_t write_rpc_req(struct bt_conn *conn, const struct bt_gatt_attr *attr, const void *buf,
+                             uint16_t len, uint16_t offset, uint8_t flags) {
+    if (handling_rx) {
+        uint32_t copied = 0;
+        struct ring_buf *rpc_buf = zmk_rpc_get_rx_buf();
+        do {
+            uint8_t *buffer;
+            uint32_t claim_len = ring_buf_put_claim(rpc_buf, &buffer, len - copied);
+
+            if (claim_len > 0) {
```

Comment:
So... It will only infinitely look *if* there's no consumer pulling out of the ring buffer. I will tweak this to be a regular while loop though to be safe.